src/umockdev.vala

```diff
@@ -745,6 +751,57 @@ public class Testbed: GLib.Object {
         this.ev_sender.send(devpath, action);
     }
 
+    /**
+     * umockdev_testbed_attach_ioctl:
+     * @self: A #UMockdevTestbed.
+     * @dev: (not nullable): Device path (/dev/...) to attach to.
+     * @handler: a #UMockdevIoctlBase instance to handle requests
+     * @error: return location for a GError, or %NULL
+     *
+     * Attach an #UMockdevIoctlBase object to handle ioctl, read and write
+     * syscalls on the given device file. This allows emulating arbitrary
+     * ioctl's and read/write calls for which umockdev does not have builtin
+     * support.
+     *
+     * Returns: %TRUE on success, %FALSE on error.
+     */
+    public bool attach_ioctl (string dev, IoctlBase handler) throws GLib.Error
+    {
+        assert (!this.custom_handlers.contains (dev));
+
+        string sockpath = Path.build_filename(this.root_dir, "ioctl", dev);
+        handler.register_path(this.worker_ctx, dev, sockpath);
+
+        this.custom_handlers.insert(dev, handler);
+
+        return true;
+    }
+
+    /**
+     * umockdev_testbed_detach_ioctl:
+     * @self: A #UMockdevTestbed.
+     * @dev: Device path (/dev/...) to detach.
+     * @error: return location for a GError, or %NULL
+     *
+     * Detach a custom #UMockdevIoctlBase object from an device node again. See
+     * umockdev_testbed_attach_ioctl(). It is an error to call this function
+     * if the path is not currently attached.
+     *
+     * Returns: %TRUE on success, %FALSE on error.
+     */
+    public bool detach_ioctl (string? dev) throws GLib.Error
```

Comment:
For symmetry, and as the code does not expect NULL, please drop the `?` here as well.